### PR TITLE
drivers: flash: w25q: Make spi_flash_wb_write/read support multiple pages

### PR DESCRIPTION
**Pull request by Olivier-ProGlove**
Until now `spi_flash_wb_write/read` functions were only supporting buffers that fit into a W25Q page (even if they were not checking the offset was page aligned).

This change allows these functions to read/write across multiple W25Q page.
This change is required as the caller of these functions are not aware of flash page size.

Signed-off-by: Olivier Martin olivier.martin@proglove.de
Signed-off-by: Johannes Hutter johannes@proglove.de
Signed-off-by: Michael Wendland michael@proglove.de

### CODE REVIEW

**Comment by Olivier-ProGlove** [comment]
@nashif Without this PR W25Q cannot be used with NVS and FCB as data are not correctly written. So it is quite high priority fix.

**Comment by Olivier-ProGlove** [comment]
PR rebased to solve merging conflict.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: drivers/flash/spi_flash_w25qxxdv.c [hunk 1]
```diff
@@ -146,24 +146,53 @@ static int spi_flash_wb_read(struct device *dev, off_t offset, void *data,
...
+						       bool enable, bool lock)
 {
 	struct spi_flash_data *const driver_data = dev->driver_data;
-	u8_t reg;
+	u8_t reg = 0U;
 	int ret;
 
-	if (offset < 0) {
-		return -ENOTSUP;
+	if (lock) {
```
Comment (baw-serafin) [inline drivers/flash/spi_flash_w25qxxdv.c]: Would it no be a lot easier to replace the k_sem with a K_mutext which is re-entrant than we don't have to worry if we lock twice we just have to unlock the same amount

**Review by nvlsianpu — APPROVED** [review]

**Review by aurel32 — APPROVED** [review]
I have no way to test this patch, but I have reviewed the code. It looks fine to me.

**Pull request merged**